### PR TITLE
TRUNK 3623 : Added ProviderDatatype,handler and tests

### PR DESCRIPTION
**Pull request by maany**
JIRA issue : https://issues.openmrs.org/browse/TRUNK-3623

### CODE REVIEW

**Comment by dkayiwa** [comment]
@maany can you squash these commits into one as advised at http://go.openmrs.org/pull-request-tips?

File: api/src/main/java/org/openmrs/customdatatype/datatype/ProviderDatatype.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import org.openmrs.customdatatype.CustomDatatype.Summary;
+import org.openmrs.customdatatype.SerializingCustomDatatype;
+import org.springframework.stereotype.Component;
+/**
+ * Datatype for a Provider, represented by {@link org.openmrs.Provider}
+ * @since 1.12
+ *
+ */
+@Component
+public class ProviderDatatype extends SerializingCustomDatatype<Provider> {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/customdatatype/datatype/ProviderDatatype.java]: The ticket description was edited, can you please extend BaseMetadataDatatype?

File: api/src/main/java/org/openmrs/customdatatype/datatype/ProviderDatatype.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	
+	/**
+	 * @see org.openmrs.customdatatype.SerializingCustomDatatype#serialize(java.lang.Object)
+	 * @should return a provider uuid during serialization
+	 */
+	@Override
+	public String serialize(Provider typedValue) {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/customdatatype/datatype/ProviderDatatype.java]: You won't need to override this

File: api/src/main/java/org/openmrs/customdatatype/datatype/ProviderDatatype.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	
+	/**
+	 * @see  org.openmrs.customdatatype.SerializingCustomDatatype#serialize(java.lang.Object)
+	 * @should	construct a Provider(org.openmrs.Provider) serialized by this handler
+	 */
+	@Override
+	public Provider deserialize(String serializedValue) {
+		if (StringUtils.isEmpty(serializedValue))
+			return null;
+		Provider provider = Context.getProviderService().getProviderByUuid(serializedValue);
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/customdatatype/datatype/ProviderDatatype.java]: You can merge these 2 lines into one with something like:

return Context.getProviderServiece.getProviderByUuid(serializedValue)

File: api/src/main/java/org/openmrs/customdatatype/datatype/ProviderDatatype.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+		return provider;
+	}
+	
+	@Override
+	public Summary getTextSummary(String referenceString) {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/customdatatype/datatype/ProviderDatatype.java]: You can get rid of this

File: api/src/main/java/org/openmrs/customdatatype/datatype/ProviderDatatype.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+		Provider provider = Context.getProviderService().getProviderByUuid(referenceString);
+		return doGetTextSummary(provider);
+	}
+	
+	@Override
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/customdatatype/datatype/ProviderDatatype.java]: And this

File: web/src/main/java/org/openmrs/web/attribute/handler/ProviderFieldGenDatatypeHandler.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+import org.openmrs.api.context.Context;
+import org.openmrs.customdatatype.CustomDatatype;
+import org.openmrs.customdatatype.CustomDatatype.Summary;
+import org.openmrs.customdatatype.InvalidCustomValueException;
+import org.openmrs.customdatatype.datatype.ProviderDatatype;
+/**
+ * An implementation of {@link FieldGenDatatypeHandler} for {@link ProviderDatatype}
+ * @since 1.12
+ *
+ */
```
Comment (wluyima) [inline web/src/main/java/org/openmrs/web/attribute/handler/ProviderFieldGenDatatypeHandler.java]: This should extend BaseMetadataFieldGenDatatypeHandler and get rid of the methods here that are already implemented in the base class

File: web/src/test/java/org/openmrs/web/attribute/handler/ProviderFieldGenDatatypeHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+	 * HttpServletRequest implementation is provided by the Servlet Container
+	 * (Tomcat/ Jetty). But for testing purpose, this local implementation is
+	 * being used as the HttpServletRequqest object is not available in the
+	 * testing environment. Please note that only the
+	 * {@link HttpServletRequest#getParameter(String)} has been defined suitably
+	 * for the purpose of this test.
+	 * 
+	 *
+	 */
+	private class HttpServletRequestImpl implements HttpServletRequest {
```
Comment (wluyima) [inline web/src/test/java/org/openmrs/web/attribute/handler/ProviderFieldGenDatatypeHandlerTest.java]: Can you please use spring's MockHttpServletRequest, you won't need this class

**Comment by dkayiwa** [comment]
Closing because author no longer works on it. 
@maany feel free to tell me reopen if you want to resume working on this. :)

**Pull request closed without merging**